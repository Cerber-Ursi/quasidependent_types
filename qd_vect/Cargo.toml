[package]
name = "qd_vect"
version = "0.1.0"
authors = ["k.bagrov <k.bagrov@g.nsu.ru>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
qd_trait = { path = "../qd_trait" }
qd_nat = { path = "../qd_nat", features = ["typenum_consts", "nat_ops", "fin"] }
typenum = "1.11"

[dev-dependencies]
trybuild = "1.0.17"
