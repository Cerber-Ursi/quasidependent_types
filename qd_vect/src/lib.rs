use qd_core::*;
use qd_nat::*;
use std::marker::PhantomData;
use std::ops::Index;

#[derive(Clone)]
pub struct Vect<T, N: Nat>(Vec<T>, PhantomData<N>);
impl<Item, N: Nat> Dependent for Vect<Item, N> {
    type Native = Vec<Item>;
    type Frozen = [Item];
    fn freeze(&self) -> &Self::Frozen {
        self.0.as_slice()
    }
    fn freeze_mut(&mut self) -> &mut Self::Frozen {
        self.0.as_mut_slice()
    }
    fn into_native(self) -> Self::Native {
        self.0
    }
    fn as_native(&self) -> &Self::Native {
        &self.0
    }
}

pub fn collect<Item, N: Nat, N2: Nat + From<N>, I: IntoIterator<Item = Item>>(
    iter: I,
) -> (N, Vect<Item, N>) {
    let inner: Vec<_> = iter.into_iter().collect();
    (expect_nat(inner.len()), Vect(inner, PhantomData))
}

#[macro_export]
macro_rules! vect {
    ($data:expr) => {
        ::qd_nat::with_n! {
            let v = $data;
            $crate::collect::<_, _, N, _>(v)
        }
    };
}

impl<Item, N: Nat> Vect<Item, N> {
    pub fn retag<New: Nat>(self, _proof: Equiv<N, New>) -> Vect<Item, New> {
        Vect(self.0, PhantomData)
    }

    pub fn size_refl(&self) -> Equiv<N, N> {
        Equiv::refl()
    }

    pub fn push(mut self, item: Item) -> Vect<Item, Add<N, typenum::consts::U1>> {
        self.0.extend(std::iter::once(item));
        Vect(self.0, PhantomData)
    }

    pub fn find_index(&self, mut pred: impl FnMut(&Item) -> bool) -> Option<Fin<N>> {
        N::iter_until().find(|&n| pred(&self[n]))
    }

    pub fn concat<N2: Nat>(mut self, vect: Vect<Item, N2>) -> Vect<Item, Add<N, N2>> {
        self.0.extend(vect.into_native());
        Vect(self.0, PhantomData)
    }
}

impl<Item, N: Nat> Index<Fin<N>> for Vect<Item, N> {
    type Output = Item;
    fn index(&self, index: Fin<N>) -> &Item {
        unsafe { self.0.get_unchecked(index.as_usize()) }
    }
}
