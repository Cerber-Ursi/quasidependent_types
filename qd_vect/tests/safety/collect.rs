use qd_vect::collect;

fn test_1() {
    let _ = collect(vec![]);
}

fn main() {}
