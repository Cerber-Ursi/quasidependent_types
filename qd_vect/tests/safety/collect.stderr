error[E0283]: type annotations needed for `(N, qd_vect::Vect<T, N>)`
  --> $DIR/collect.rs:4:13
   |
4  |     let _ = collect(vec![]);
   |         -   ^^^^^^^
   |         |   |
   |         |   cannot infer type for type parameter `N` declared on the function `collect`
   |         |   help: consider specifying the type arguments in the function call: `collect::<Item, N, I>`
   |         consider giving this pattern the explicit type `(N, qd_vect::Vect<T, N>)`, where the type parameter `N` is specified
   | 
  ::: $WORKSPACE/qd_vect/src/lib.rs:25:25
   |
25 | pub fn collect<Item, N: Nat, I: IntoIterator<Item = Item>>(iter: I) -> (N, Vect<Item, N>) {
   |                         --- required by this bound in `qd_vect::collect`
   |
   = note: cannot resolve `_: qd_nat::timestamped::nat::Nat`
